src/sas/sascalc/calculator/geni.py

```diff
@@ -510,4 +512,109 @@ def _spin_weights(in_spin, out_spin):
     )
     return weight
 
+def radius_of_gyration(nuc_sld_data):
+    #Calculate Center of Mass(CoM) First
+    CoM = centerOfMass(nuc_sld_data)
+
+    #Now Calculate RoG
+    guinier_num = guinier_den = rog_num = rog_den = 0.0
+
+    dtype = np.dtype('double')
+    x, y, z = nuc_sld_data.pos_x, nuc_sld_data.pos_y, nuc_sld_data.pos_z
+    pix_symbol = nuc_sld_data.pix_symbol
+    coordinates = np.array([x, y, z], dtype=dtype).T
+    slds, masses = np.empty(len(pix_symbol), dtype), np.empty(len(pix_symbol), dtype)
+    for i, sym in enumerate(pix_symbol):
+        atom = periodictable.elements.symbol(sym)
+        masses[i] = atom.mass
+        slds[i] = atom.neutron.b_c
+        #solvent_slds = atoms.volume() * 10**24 * float(self.txtSolventSLD.text()) * 10**5
+
+    #TODO: Implement a scientifically sound method for obtaining protein volume - Current value is a inprecise approximation. Until then Solvent SLD does not impact RG - SLD.
+    # contrastSLD = sld - solvent_sld         #femtometer
+    contrastSLDs = slds                       #femtometer
+    rsq = np.sum((CoM - coordinates)**2, axis=1)
+
+    rog_num = np.sum(masses * rsq)
+    rog_den = np.sum(masses)
+    guinier_num = np.sum(contrastSLDs * rsq)
+    guinier_den = np.sum(contrastSLDs)
+
+
+    if rog_den <= 0: #Should never happen as there are no zero or negative mass atoms
+        rog_mass = "NaN"
+        logging.warning("Atomic Mass is zero for all atoms in the system.")
+    else:
+        rGMass = np.sqrt(rog_num/rog_den)
+        rog_mass = (str(round(np.sqrt(rog_num/rog_den),1)) + " Å")
+
+    #Avoid division by zero - May occur through contrast matching
+    if guinier_den == 0:
+        guinier_value = "NaN"
+        logging.warning("Effective Coherent Scattering Length is zero for all atoms in the system.")
+    elif (guinier_num/guinier_den) < 0:
+        guinier_value = (str(round(np.sqrt(-guinier_num/guinier_den),1)) + " Å")
+        logging.warning("Radius of Gyration Squared is negative. R(G)^2 is assumed to be |R(G)|* R(G).")
+    else:
+        guinier_value = (str(round(np.sqrt(guinier_num/guinier_den),1)) + " Å")
+
+    return [rog_mass,guinier_value,rGMass]          #[String, String, Float], float used for plugin model
```

Comment:
Don't rely on position to give meaning to variables, use a structured datatype, preferably a dataclass, or at a stretch, a named tuple